### DISCUSSION TITLE
Some .CBR files open extremely slower than others

### DISCUSSION THREAD

**Discussion by wjharing** [sumatrapdfreader/sumatrapdf#4561] [General]
My comics are in .CBR/.CBZ file format.
I noticed that some .CBR/.CBZ files open a lot slower and it looks as if program is hanging (for 12 seconds!), before it opens correctly.

An example of such file is 70mb, so that shouldn't be the issue.
Some files might have subfolders in them, before the pages are listed. This shouldn't be a problem either I believe.

I did notice with one slow .CBR file, that the pages are numbered 001.jpg, etc, but that one of the page numbers is omitted. Would this be a reason why SumatraPDFReader struggles?

**Comment by GitHubRulesOK** [comment]
Rendering timings can vary considerably from version to version as the core code goes through many altered balance of demands on resources. So the only litmus test is current pre-release as that can be changed for any specific variability in behaviours.

Fastest time will usually only be achieved with local file on main (C:) drive.
However just tested a 19.5 MB (623 pages!) on a remote device L:\ and it appeared in about 3 seconds 
allowing for transportation the log shows it was about 1 second to load.
```
LoadDocument: 43.77 ms, 623 pages for 'L:\!fromHP\Desktop\temp\continued-fractions.djvu_0623.cbz'
DisplayModel::BuildPagesInfo started
DisplayModel::BuildPagesInfo took 858.14 ms
EngineCbx::LoadBitmapForPage(page: 2) took 0.65 ms
```


However in order for there to be a change specific example files that demonstrate such lagging would be needed and GitHub imposes size restrictions.

**Comment by wjharing** [comment]
Sorry, I should have clarified, I was using the last pre-release version of 3.6. (but noticed with many previous 3.6 versions as well)

Log for accessing the .cbr that works normally:
```
CreateControllerForEngineOrFile: 'C:\!ncoming\2 test sumatra\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr', 50 pages
uitask::WndProcTaskDispatch: will execute 'TaskLoadDocumentAsyncFinish', func 0x000001EF2C3C18E0
DisplayModel::BuildPagesInfo started
DisplayModel::BuildPagesInfo took 449.14 ms
EngineCbx::LoadBitmapForPage(page: 2) took 26.37 ms
Slow rendering: 109.62 ms, page: 2 in 'C:\!ncoming\2 test sumatra\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 1) took 3.27 ms
```

The log for accessing the .cbr that is opening with a 12s delay (freezing up):
```
CreateControllerForEngineOrFile: 'C:\!ncoming\2 test sumatra\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr', 49 pages
uitask::WndProcTaskDispatch: will execute 'TaskLoadDocumentAsyncFinish', func 0x000001EF2C3C2BE0
DisplayModel::BuildPagesInfo started
DisplayModel::BuildPagesInfo took 13610.73 ms
EngineCbx::LoadBitmapForPage(page: 2) took 64.12 ms
Slow rendering: 167.35 ms, page: 2 in 'C:\!ncoming\2 test sumatra\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 1) took 42.10 ms
Slow rendering: 151.69 ms, page: 1 in 'C:\!ncoming\2 test sumatra\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
```

Hereby a link to both files.
[https://gofile.io/d/FZh98J](url)

**Comment by GitHubRulesOK** [comment]
Hmmm I must agree that using those files in different ways SumatraPDF seem slower however converted to PDF they seem much faster ! the file is larger as PDF but the speed seems good so for what it is worth that would be a workaround until pre-release fares better you can even use older portable SumatraPDF to do the job. https://filetransfer.io/data-package/PMKzbWAV#link
image.pdf was a test with other software for comparison and the smaller one is the pure cbr saved as PDF from version 3.1.2

When source images are jpeg then pdf can hold them "as is" with very little overheads as page objects

![image](https://github.com/user-attachments/assets/8723a241-96c8-44aa-b37f-61a17f08b908)

**Comment by wjharing** [comment]
When I unpack and repack that archive with the problem, the problem disappears.
rar info (unrar lt <comic.cbr>) from the 01 comic file (the file with the issue) shows RAR 1.5(v29) -m3 -md=**4m**
While the .rar info from the 02 comic shows RAR 1.5(v29) -m3 -md=2m
And when I repack, RAR 5.0(v50) -m3 -md=2m

I don't know if -md=**4m** is causing issues, given this has to do with memory, compression, cpu? 


I checked your generated .pdfs, and those are rendering quick. But indeed a temporary workaround solution only. Hopefully the above info helps in any way?

**Comment by kjk** [comment]
the file is no longer available

**Reply by GitHubRulesOK** [reply]
@kjk 
sorry have been at heavens door in hospital but they refused to keep me in , Cant be certain this was the file state at that time (had reduced to 049.cbz while OP said 50 pages), but should be similar enough the PDF was a trial run of mine the 049.cbz was from the OP source file.
Sorry  Link has now timed out !

**Reply by GitHubRulesOK** [reply]
@kjk 
OK  think I tracked them down so same link as above there are 2 CBR files that were the likely source.

**Reply by wjharing** [reply]
It seems you have the files, but let me know if you need anything further to help identify the issue

**Comment by wjharing** [comment]
ps, I don't know what unrar/unzip engines are used within sumatrapdf in order to unpack the .cbr or .cbz files, and if upgrading to the latest version would address this particular issue of slow opening times for select comic files?
If there is anything further I can provide, to help resolve this, let me know.

**Reply by GitHubRulesOK** [reply]
Developer will have tested several options for speed see several more recent commits in November related to de/compression.
My links to samples are time deleted so not accessible to @kjk is there a way you can provide a longer term link or a smaller file here with similar behaviours.

AFAIK it was just the fact that the images were higher density but will be affected by de/ compression strategy. PDF are often faster as less compressed.

**Reply by wjharing** [reply]
Thanks for looking into this.

I have shared the files again, with longer expiry time (< 31/Mar):
https://gofile.io/d/S9abDx

The 01 file, is the one with the very long opening time, and the 02 file opens quickly. Both similar sized files, but from what I can see, the archives are with different compression tags)

Hopefully this helps further to fix this quirky issue

**Comment by GitHubRulesOK** [comment]
Hmm there does appear to be a clear difference the render is much slower as there is a page size difference, but the loading times seem to be increasing noticeably per page rather than roughly the same?

File: C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr
File Size: 68.27 MB (71,587,920 Bytes)
Number of Pages: 49
Page Size: 54.74 x 75.22 cm

File: C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr
File Size: 49.24 MB (51,629,669 Bytes)
Number of Pages: 50
Page Size: 44.98 x 61.89 cm

```
uitask::WndProcTaskDispatch: will execute 'SafeDeleteCommandPaletteWnd', func 0x0393A1F8
uitask::WndProcTaskDispatch: did execute, will delete func 0x0393A1F8
HandleExecuteCmds: '[Open("C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr", 0, 1, 0)]'
HandleOpenCmd: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr', newWindow: 0, setFocus: 1, forceRefresh: 0, inCurrentTab: 0, isCtrl: 0
HandleOpenCmd: using the only window
HandleOpenCmd: calling LoadDocument(), activateExisting: 1
EngineCbx::CreateFromFile(): opening archive took 3.67
EngineCbx::FinisHLoading() in 16.69 ms
CreateControllerForEngineOrFile: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr', 49 pages, took 21 ms
LoadDocument: 20.57 ms, 49 pages for 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
DisplayModel::BuildPagesInfo started
DisplayModel::BuildPagesInfo took 22835.57 ms
CreateThumbnailForFile: filePath: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr', 0x03926538, d: 0x0393A110
SaveSettings
EngineCbx::LoadBitmapForPage(page: 2) took 157.35 ms
NotifyAboutFile: ignoring 'C:\Users\K\AppData\Local\SumatraPDF\SumatraPDF-settings.txt'
NotifyAboutFile: ignoring 'C:\Users\K\AppData\Local\SumatraPDF\SumatraPDF-settings.txt'
HandleOpenCmd: setFocus != 0 so calling win->Focus()
EngineImages::RenderPage() in 393.30 ms
Slow rendering: 550.83 ms, page: 2 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 1) took 120.28 ms
Slow rendering: 195.65 ms, page: 1 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
CreateThumbnailFinish: path: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr', 0x03926538, d: 0x0393A110, d->bmp: 0x03939128
~CreateThumbnailData: deleting 0x0393A110 filePath='C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr' 0x03926538
Slow rendering: 146.42 ms, page: 1 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
currTabIdx: 5
Selection changed: 0
Selection changed: 0
Selection changed: 0
Selection changed: 0
uitask::WndProcTaskDispatch: will execute 'SafeDeleteCommandPaletteWnd', func 0x018684A8
uitask::WndProcTaskDispatch: did execute, will delete func 0x018684A8
EngineCbx::LoadBitmapForPage(page: 3) took 153.98 ms
Slow rendering: 278.49 ms, page: 3 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 4) took 155.60 ms
Slow rendering: 278.89 ms, page: 4 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 5) took 179.39 ms
Slow rendering: 346.28 ms, page: 5 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 6) took 183.84 ms
Slow rendering: 305.61 ms, page: 6 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 7) took 212.28 ms
Slow rendering: 333.14 ms, page: 7 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 8) took 231.33 ms
Slow rendering: 351.62 ms, page: 8 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 9) took 244.29 ms
Slow rendering: 365.52 ms, page: 9 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 10) took 258.70 ms
Slow rendering: 380.14 ms, page: 10 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 11) took 273.55 ms
Slow rendering: 388.63 ms, page: 11 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 12) took 288.12 ms
Slow rendering: 402.30 ms, page: 12 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 01 - De gruwel in de storm.cbr'
EngineCbx::LoadBitmapForPage(page: 13) took 307.15 ms
```

```
HandleExecuteCmds: '[Open("C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr", 0, 1, 0)]'
HandleOpenCmd: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr', newWindow: 0, setFocus: 1, forceRefresh: 0, inCurrentTab: 0, isCtrl: 0
HandleOpenCmd: using the only window
HandleOpenCmd: calling LoadDocument(), activateExisting: 1
EngineCbx::CreateFromFile(): opening archive took 247.60
EngineCbx::FinisHLoading() in 18.67 ms
CreateControllerForEngineOrFile: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr', 50 pages, took 289 ms
LoadDocument: 289.22 ms, 50 pages for 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
DisplayModel::BuildPagesInfo started
DisplayModel::BuildPagesInfo took 3639.54 ms
StartMonitoringDirForChangesAPC() C:\Users\K\Downloads
CreateThumbnailForFile: filePath: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr', 0x03819748, d: 0x01868100
SaveSettings
NotifyAboutFile: ignoring 'C:\Users\K\AppData\Local\SumatraPDF\SumatraPDF-settings.txt'
NotifyAboutFile: ignoring 'C:\Users\K\AppData\Local\SumatraPDF\SumatraPDF-settings.txt'
EngineCbx::LoadBitmapForPage(page: 2) took 122.55 ms
Slow rendering: 219.25 ms, page: 2 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 1) took 7.07 ms
CreateThumbnailFinish: path: 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr', 0x03819748, d: 0x01868100, d->bmp: 0x038D5A28
~CreateThumbnailData: deleting 0x01868100 filePath='C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr' 0x03819748
HandleOpenCmd: setFocus != 0 so calling win->Focus()
EngineCbx::LoadBitmapForPage(page: 3) took 93.35 ms
Slow rendering: 187.64 ms, page: 3 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 6) took 71.87 ms
Slow rendering: 166.09 ms, page: 6 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 4) took 65.72 ms
Slow rendering: 160.97 ms, page: 4 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 5) took 71.65 ms
Slow rendering: 170.64 ms, page: 5 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 8) took 61.78 ms
Slow rendering: 152.21 ms, page: 8 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 9) took 76.35 ms
Slow rendering: 167.64 ms, page: 9 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 7) took 64.23 ms
Slow rendering: 159.93 ms, page: 7 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 10) took 75.26 ms
Slow rendering: 165.21 ms, page: 10 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 12) took 79.42 ms
Slow rendering: 165.90 ms, page: 12 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 11) took 80.04 ms
Slow rendering: 172.27 ms, page: 11 in 'C:\Users\K\Downloads\20.000 Eeuwen onder zee - 02 - Het schuilhol van Cthulhu.cbr'
EngineCbx::LoadBitmapForPage(page: 14) took 81.60 ms
```

**Reply by wjharing** [reply]
When the slow file is repacked, the issue seems to be resolved, so it seems that the issue is not with the images themselves, but just the .rar container, that in certain compression configurations seems to throw a fit.

**Reply by wjharing** [reply]
I noticed that my files have been auto-removed from gofile, but let me know if I have to reupload.

Hoping a new version of sumatrapdf, will make some speed improvements for those specific cbr files. 👍

**Comment by danielmmmm** [comment]
Do any of you load those slow files from a network drive by any chance? I just opened a bug report for slow loading from network drives (https://github.com/sumatrapdfreader/sumatrapdf/issues/5757).

**Comment by kjk** [comment]
In latest [pre-release](https://www.sumatrapdfreader.org/prerelease) loading comic book files has been greatly improved
